[dynamo, nested graph breaks] move cell codegen before side effects codegen

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #160611
* #163335
* #163503
* #162737
* __->__ #160601

This is needed because if we codegen cells for nested frames AFTER side effects, then reconstruction could get messed up. From below:

>The added test case demonstrates the reconstruction failure if we kept cell codegen at the original place (only happens with nested graph breaks since we reconstruct nested frame cells from VariableTracker rather than directly using LOAD_CLOSURE).

>At a high level, what happened before this change was that side_effects was pruning the cells (I don't recall exactly why this happens), and because cells were codegen'd after the side effects were applied, we were unable to properly reconstruct the cell. The error I was seeing was a list/tuple IndexError.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

williamwen42 committed:
[dynamo, nested graph breaks] move cell codegen before side effects codegen

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] move cell codegen before side effects codegen"




[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] move cell codegen before side effects codegen"




[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] move cell codegen before side effects codegen"




[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] move cell codegen before side effects codegen"




[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] move cell codegen before side effects codegen"




[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] move cell codegen before side effects codegen"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] move cell codegen before side effects codegen"


This is needed because if we codegen cells for nested frames AFTER side effects, then reconstruction could get messed up.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] move cell codegen before side effects codegen"


This is needed because if we codegen cells for nested frames AFTER side effects, then reconstruction could get messed up.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] move cell codegen before side effects codegen"


This is needed because if we codegen cells for nested frames AFTER side effects, then reconstruction could get messed up.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

Lucaskabela reviewed (changes requested):
Seems the flexAttention tests are failing, although from the error message I am not sure if that is from this PR or not - can you rebase?

Lucaskabela commented on torch/_dynamo/variables/iter.py:
```diff
@@ -20,7 +20,11 @@
 
 from .. import graph_break_hints, polyfills, variables
-from ..bytecode_transformation import create_call_function, create_instruction
+from ..bytecode_transformation import (
```
Comment for copying

williamwen42 committed:
Update on "[dynamo, nested graph breaks] move cell codegen before side effects codegen"


This is needed because if we codegen cells for nested frames AFTER side effects, then reconstruction could get messed up.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

mlazos commented on torch/_dynamo/output_graph.py:
```diff
@@ -1522,7 +1523,7 @@ def compile_subgraph(
         # see symbolic_convert.py for
         # codegen stack convention after the unsupported instruction
-        # NOTE: cells are loaded into continuation functions directly
+        # NOTE: cells will loaded into continuation functions directly by symbolic_convert
```
nit: typo

mlazos reviewed (approved):
Looks good, can you add an explicit example in the PR desc when reconstruction will get messed up?

williamwen42 commented:
The added test case demonstrates the reconstruction failure if we kept cell codegen at the original place (only happens with nested graph breaks since we reconstruct nested frame cells from VariableTracker rather than directly using LOAD_CLOSURE).

At a high level, what happened before this change was that side_effects was pruning the cells (I don't recall exactly why this happens), and because cells were codegen'd after the side effects were applied, we were unable to properly reconstruct the cell. The error I was seeing was a list/tuple IndexError.

williamwen42 committed:
Update on "[dynamo, nested graph breaks] move cell codegen before side effects codegen"


This is needed because if we codegen cells for nested frames AFTER side effects, then reconstruction could get messed up. From below:

>The added test case demonstrates the reconstruction failure if we kept cell codegen at the original place (only happens with nested graph breaks since we reconstruct nested frame cells from VariableTracker rather than directly using LOAD_CLOSURE).

>At a high level, what happened before this change was that side_effects was pruning the cells (I don't recall exactly why this happens), and because cells were codegen'd after the side effects were applied, we were unable to properly reconstruct the cell. The error I was seeing was a list/tuple IndexError.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo, nested graph breaks] move cell codegen before side effects codegen"


This is needed because if we codegen cells for nested frames AFTER side effects, then reconstruction could get messed up. From below:

>The added test case demonstrates the reconstruction failure if we kept cell codegen at the original place (only happens with nested graph breaks since we reconstruct nested frame cells from VariableTracker rather than directly using LOAD_CLOSURE).

>At a high level, what happened before this change was that side_effects was pruning the cells (I don't recall exactly why this happens), and because cells were codegen'd after the side effects were applied, we were unable to properly reconstruct the cell. The error I was seeing was a list/tuple IndexError.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]
